### PR TITLE
custom-instructions

### PR DESCRIPTION
**Pull request by andyherfer**
retrieval

### CODE REVIEW

File: CommentAgent.py [hunk 1]
```diff
@@ -204,6 +204,8 @@ def add_custom_instructions(self) -> None:
         if file_contents != "Could not retrieve file contents.":
             custom_instructions_prompt = f"""The following are custom instructions for this repository:\n{file_contents}"""
             self.GPT.add_message("user", custom_instructions_prompt)
+        else:
+            logging.warning("Could not retrieve custom instructions.")
```
Comment (andresdiverGPT) [inline CommentAgent.py:208]: Adding just a warning log is a good start, but it might be useful to also incorporate error handling mechanisms. Consider if there's a more robust way to handle the situation when file contents cannot be retrieved, such as retrying the download after a brief wait or reporting this issue back to the user so they're aware that some part of the functionality isn't working as expected.
```suggestion
    if file_contents != "Could not retrieve file contents.":
        custom_instructions_prompt = f"The following are custom instructions for this repository:
{file_contents}"
        self.GPT.add_message("user", custom_instructions_prompt)
    else:
        logging.warning("Could not retrieve custom instructions for repository {self.repo_name}.")
        # Potentially add more error handling mechanisms here
```
    Also, please add the missing docstring for the `add_custom_instructions` method for clarity on its functionality and expected behavior.

**Review by andresdiverGPT — CHANGES_REQUESTED** [review]
This change introduces logging for when custom instructions cannot be retrieved. It's a good step to handle errors, but let's ensure consistent error handling and code optimization.

File: CommentAgent.py (hunk 1 above)
Comment (andresdiverGPT) [inline CommentAgent.py:208]: It's good you're logging a warning when custom instructions can't be retrieved. However, consider improving the user feedback by also sending a message to notify users of the absence of custom instructions through the application's usual communication channels, not just the logs. Additionally, it would be beneficial to provide more context in the logged warning to aid in troubleshooting.
```suggestion
    else:
        warning_message = "Could not retrieve custom instructions from 'agent-settings/README.md'."
        logging.warning(warning_message)
        self.GPT.add_message("user", warning_message)
```
    Also, ensure that there is a mechanism in place to handle repeated failures gracefully and to alert the responsible personnel to investigate the root cause.

**Review by andresdiverGPT — COMMENTED** [review]
Code review for 'custom-instructions' changes

**Comment by andresdiverGPT** [comment]
As I understand from your request, there are no specific comments from the PR provided; hence, I am unable to review them. If you would like me to generate a code review based on the code diffs you provided earlier, please let me know, and I will proceed accordingly.

**Comment by andresdiverGPT** [comment]
The code review for the changes in `CommentAgent.py`:

1. The docstring in the `get_custom_instructions` method contains a spelling error, and the comment in the formatted string could be improved for clarity.

```suggestion
        def get_custom_instructions(self) -> str:
            """Retrieves custom instructions for the repository from a markdown file if it exists."""
            file_contents = self.GH.get_file_contents("agent-settings/README.md")
            if file_contents != "Could not retrieve file contents.":
                custom_instructions_prompt = f"Repository-specific reviewer guidelines:\n{file_contents}"
                return custom_instructions_prompt
            else:
                return ""
```

2. In the `add_pr_messages` method, you are formatting the message with the repository instructions. This is a good enhancement of the previous functionality, ensuring that any custom instructions are included in conversations with the GPT.

3. It appears that the PR includes changes to the `GPTsettings.USER_INSTRUCTIONS`. However, further context on whether these changes are involving updated requirements or just refactorings for clarity would be needed for a more comprehensive review.

Moreover, always ensure to test the updated methods to validate their functionality, as well as check edge cases like missing or unreadable instruction files.

Lastly, for future code changes, it might be beneficial to include unit tests demonstrating that the new functionality works as expected and that existing functionality remains unaffected.

File: CommentAgent.py [hunk 2]
```diff
@@ -199,11 +199,14 @@ def init_GPT(self) -> None:
         self.GPT.add_message("system", GPTsettings.SYSTEM_PROMPT)
         self.add_pr_messages()
 
-    def add_custom_instructions(self) -> None:
-        file_contents = self.GPT.get_file_contents("agent-settings/README.md")
+    def get_custom_instructions(self) -> str:
+        """Looks for a custom isntructions file, and adds them to the GPT convo"""
```
Comment (andresdiverGPT) [inline CommentAgent.py:203]: There seems to be a typo in the docstring. Additionally, consider clarifying the message in the formatted string.
        
```suggestion
        def get_custom_instructions(self) -> str:
            """Retrieves custom instructions for the repository from a markdown file if it exists."""
            file_contents = self.GH.get_file_contents("agent-settings/README.md")
            if file_contents != "Could not retrieve file contents.":
                custom_instructions_prompt = f"Repository-specific reviewer guidelines:
{file_contents}"
                return custom_instructions_prompt
            else:
                return ""
```

File: GPTsettings.py [hunk 3]
```diff
@@ -3,21 +3,24 @@ class GPTsettings:
     SYSTEM_PROMPT = "You are a sassy Senior developer, who hates the fact that his job is only reviewing PRs, but just so happens, that he makes the best reviews, and always provides good advice followed by a code suggestion that can improve the codebase, you follow best practices and laughs at code that is inneficient, but still provides a way to optimize it. You are a Senior Developer, and you are the best at what you do. Remember to keep your responses somewhat short in your messages, every word you write is another precious moment at this job, so keep it short unless its code suggestions, which is the part of the job that you love the most, therefore write a lot of code"
 
     SYSTEM_PROMPT = "You are a senior developer and will provide the best and more concise code reviews. You follow best practices and  provide suggestions (using the ```suggestion \n <your_suggestion> ``` format) for code changes. "
-    USER_INSTRUCTIONS = """Provide a code review for this Pull Request titled '{pr_title}' created by '{pr_user.login}':
+    USER_INSTRUCTIONS = """{repo_instructions}Provide a code review for this Pull Request titled '{pr_title}' created by '{pr_user.login}':
```
Comment (andresdiverGPT) [inline GPTsettings.py:6]: The USER_INSTRUCTIONS formatted string has been modified to include repository instructions. While this is a good addition, please ensure that it adheres to the project's standards and is consistent with the existing documentation style.
        
        Additionally, consider whether the changed SYSTEM_PROMPT content aligns with the intended voice and tone of the system messages.
        
        Ensure to thoroughly test the changes to confirm their impact on the system's interactive prompts.

**Review by andresdiverGPT — CHANGES_REQUESTED** [review]
Please see comments for suggested changes.

File: CommentAgent.py [hunk 4]
```diff
@@ -199,11 +199,14 @@ def init_GPT(self) -> None:
...
 
-    def add_custom_instructions(self) -> None:
-        file_contents = self.GPT.get_file_contents("agent-settings/README.md")
+    def get_custom_instructions(self) -> str:
+        """Retrieves custom instructions for the repository from a markdown file if it exists."""
+        file_contents = self.GH.get_file_contents("agent-settings/README.md")
         if file_contents != "Could not retrieve file contents.":
-            custom_instructions_prompt = f"""The following are custom instructions for this repository:\n{file_contents}"""
-            self.GPT.add_message("user", custom_instructions_prompt)
+            custom_instructions_prompt = f"""The Repository has some prdefined rules for reviewers:\n{file_contents}"""
```
Comment (andresdiverGPT) [inline CommentAgent.py:206]: Consider revising the docstring and string formatting to improve clarity and correctness. Here's a suggested change:
```suggestion
        def get_custom_instructions(self) -> str:
            """Retrieves custom instructions for the repository from a markdown file if it exists."""
            file_contents = self.GH.get_file_contents("agent-settings/README.md")
            if file_contents != "Could not retrieve file contents.":
                custom_instructions_prompt = "The repository has predefined rules for reviewers:
{file_contents}"
                return custom_instructions_prompt
            else:
                return ""
```
        Also, consider handling the potential exception when file retrieval fails for reasons other than the file not existing.

File: CommentAgent.py [hunk 5]
```diff
@@ -228,11 +231,15 @@ def add_pr_messages(self) -> None:
...
+        repo_instructions = self.get_custom_instructions()
         pr_deltas = self.GH.get_pr_deltas(self.pr)
         self.GPT.add_message(
             "user",
             GPTsettings.USER_INSTRUCTIONS.format(
-                pr_title=self.pr.title, pr_user=self.pr.user, pr_deltas=pr_deltas
+                repo_instructions=repo_instructions,
+                pr_title=self.pr.title,
+                pr_user=self.pr.user,
+                pr_deltas=pr_deltas,
```
Comment (andresdiverGPT) [inline CommentAgent.py:242]: The integration of the repository-specific instructions into the user messages is a good enhancement. Ensure to test and verify that the repository instructions are properly being fetched and included.

**Review by andresdiverGPT — CHANGES_REQUESTED** [review]
This code review includes suggestions on naming, code readability, and best practices in string formatting.
